pyteal/compiler/optimizer/optimizer.py

```diff
@@ -0,0 +1,89 @@
+from typing import cast, Set
+from pyteal.ir.tealblock import TealBlock
+
+from pyteal.ir.tealop import TealOp
+from pyteal.ir.ops import Op
+
+
+class OptimizeOptions:
+    """An object which specifies the optimizations to be performed and relevant context.
+    Args:
+
+        scratch_slots (optional): cancel contiguous store/load operations
+            that have no load dependencies elsewhere.
+        skip_ids (optional): the slot ids that should be skipped during
+            optimization. At the moment this includes:
+                1. reserved slots because they may have dependencies outside
+                the current application. For example, the 'gloads' opcode can
+                access the slots of other applications in the tx group.
+                2. global slots because they're outside the scope of global
+                optimizations, which only apply to the control flow graph of
+                a single subroutine.
+                3. slots used with dynamic scratch vars. These slots use
+                indirection by means of the 'stores' opcode and dependencies
+                can only be determined at runtime."""
+
+    def __init__(self, *, scratch_slots: bool = False, skip_ids: Set[int] = None):
+        self.scratch_slots = scratch_slots
+        self.skip_ids: Set[int] = skip_ids if skip_ids is not None else set()
+
+
+def _remove_extraneous_slot_access(start: TealBlock, remove: Set[int]):
+    def keep_op(op: TealOp):
+        if type(op) != TealOp or (op.op != Op.store and op.op != Op.load):
+            return True
+
+        return cast(int, op.args[0]) not in remove
+
+    for block in TealBlock.Iterate(start):
+        block.ops = list(filter(keep_op, block.ops))
+
+
+def _has_load_dependencies(cur_block: TealBlock, start: TealBlock, slot: int, pos: int):
+    for block in TealBlock.Iterate(start):
+        for i, op in enumerate(block.ops):
+            if block == cur_block and i == pos:
+                continue
+
+            if type(op) == TealOp and op.op == Op.load and op.args[0] == slot:
+                return True
```

Comment:
If I understand it correctly, this function checks the existence of load from `slot` other than specific block and operation position, right? 

We do not distinguish before/after execution of the `load` we are ignoring, am I correctly understanding it?

Reply:
Right, this is basically dumb, overly eager dependency checking. I'll add a comment to the function for clarification.